modules/gateway/peers.go

```diff
@@ -173,20 +174,104 @@ func (g *Gateway) threadedAcceptConn(conn net.Conn) {
 	defer g.threads.Done()
 
 	addr := modules.NetAddress(conn.RemoteAddr().String())
-	g.log.Printf("INFO: %v wants to connect", addr)
+	g.log.Debugf("INFO: %v wants to connect", addr)
 
 	remoteVersion, err := acceptConnVersionHandshake(conn, build.Version)
 	if err != nil {
 		g.log.Debugf("INFO: %v wanted to connect but version handshake failed: %v", addr, err)
-		// TODO: should this be muxado.Server(conn).Close()?
 		conn.Close()
 		return
 	}
 
+	switch {
+	case build.VersionCmp(remoteVersion, "0.6.1") < 0:
+		err = g.managedAcceptConnOldPeer(conn, remoteVersion)
+	default:
+		err = g.managedAcceptConnNewPeer(conn, remoteVersion)
+	}
+	if err != nil {
+		g.log.Debugf("INFO: %v wanted to connect, but failed: %v", addr, err)
+		conn.Close()
+		return
+	}
+
+	g.log.Debugf("INFO: accepted connection from new peer %v (v%v)", addr, remoteVersion)
+}
+
+// managedAcceptConnOldPeer accepts a connection request from peers < v0.6.1.
+// The requesting peer is added as a peer, but is not added to the node list
+// (older peers do not share their dialback address). The peer is only added if
+// a nil error is returned.
+func (g *Gateway) managedAcceptConnOldPeer(conn net.Conn, remoteVersion string) error {
+	addr := modules.NetAddress(conn.RemoteAddr().String())
+
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	// Kick peers to make room if necessary.
+	g.kickPeersToMakeRoomFor(addr)
```

Comment:
definitely weird. I think you lose important context when you factor it out into a helper function. Maybe have a `g.acceptPeer` or similar that does both the `kick` and the `addPeer`, since they are common to both the old and new `managedAcceptConn`.